pkg/policy/mapstate.go

```diff
@@ -777,30 +792,66 @@ func (ms *mapState) overrideAuthRequirement(newEntry mapStateEntry, k Key, v map
 // This function is expected to be called for a map insertion after deny
 // entry evaluation. If there is a covering map key for 'newKey'
 // which denies traffic matching 'newKey', then this function should not be called.
-func (ms *mapState) authPreferredInsert(newKey Key, newEntry mapStateEntry, changes ChangeState) {
-	if !newEntry.AuthRequirement.IsExplicit() {
-		// New entry has a default auth type.
+func (ms *mapState) authPreferredInsert(newKey Key, newEntry mapStateEntry, features policyFeatures, changes ChangeState) {
+	// Bail if covered by a deny key or a key with a higher proxy port priority and current
+	// entry has no explicit auth.
+	var derived bool
+	newEntryHasExplicitAuth := newEntry.AuthRequirement.IsExplicit()
+	for k, v := range ms.CoveringKeysWithSameID(newKey) {
+		if v.IsDeny() {
+			return // bail if covered by deny
+		}
+		if v.ProxyPortPriority > newEntry.ProxyPortPriority {
+			if !newEntryHasExplicitAuth || v.AuthRequirement == newEntry.AuthRequirement {
+				// Covering entry has higher proxy port priority and newEntry has a
+				// default auth type or the same auth requirement => can bail out
+				return
+			}
+
+			// newEnry has a different explicit auth requirement, must propagate
+			// proxy port and priority and keep it
+			newEntry.ProxyPort = v.ProxyPort
+			newEntry.ProxyPortPriority = v.ProxyPortPriority
 
+			// Can break out:
+			// - if there were covering denies the allow 'v' would
+			//   not have existed, and
+			// - since the new entry has explicit auth it does not need to be
+			//   derived.
+			break
```

Comment:
What if there are two covering entries with proxy ports? Then it would not be safe to bail out here, we would need to "inherit" twice?

Reply:
In that case the most specific covering entry would have the highest priority proxy port in it, so there is no need to look at the less specific covering entries.

Reply:
Aha, missed that `CoveringKeysWithSameID` iterates most-to-least-specific.